src/deploy/functions/prepare.ts

```diff
@@ -51,10 +80,21 @@ export async function prepare(
   const firebaseConfig = await functionsConfig.getFirebaseConfig(options);
   context.firebaseConfig = firebaseConfig;
   const runtimeConfig = await getFunctionsConfig(context);
-  const env = await getEnvs(context);
+
+  utils.assertDefined(
+    options.config.src.functions.source,
+    "Error: 'functions.source' is not defined"
+  );
+  const source = options.config.src.functions.source;
+  const envs = getEnvs({
+    firebaseConfig: runtimeConfig,
+    functionsSource: options.config.path(source),
+    projectId: projectId,
+    projectAlias: options.projectAlias,
+  });
 
   logger.debug(`Analyzing ${runtimeDelegate.name} backend spec`);
-  const wantBackend = await runtimeDelegate.discoverSpec(runtimeConfig, env);
+  const wantBackend = await runtimeDelegate.discoverSpec(runtimeConfig, envs);
```

Comment:
Do we want the specified env values to be available during discovery? @inlined

Reply:
Let's start with "no" and see how much pushback we get. I'm nervous that allowing Env variables will mean that we've regressed to where parameter discovery is dynamic again.

Reply:
Oh I thought having env vars available during trigger discovery was something that we wanted to support. E.g.

```js
functions.runWith({ minInstances: process.env.DEPLOY_ENV === "prod" ? 5 : 0 })...
```

Are we saying we don't want this anymore?

Reply:
Offline - no won't support this usecase.